proxy/src/control/discovery.rs

```diff
@@ -55,8 +61,27 @@ pub struct DiscoveryWork<T: HttpService<ResponseBody = RecvBody>> {
     rx: mpsc::UnboundedReceiver<(FullyQualifiedAuthority, mpsc::UnboundedSender<Update>)>,
 }
 
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct DstLabels {
+    set: Arc<HashMap<String, String>>,
+    addr: Arc<HashMap<String, String>>,
+}
+
+/// A service that adds a label extension to all requests transiting it.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct LabelRequest<T> {
+    labels: Option<DstLabels>,
+    inner: T,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+pub struct Labeled<T> {
```

Comment:
Hmm. These types were originally written with different _intentions_, one of them annotating an entry in a `DestinationSet` with labels and the other serving as a service that adds a label to a request. However, since you mention it, it occurs to me that the same type could probably be used for both purposes --- that might simplify things here a bit?